src/shims/io.rs

```diff
@@ -123,38 +144,69 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
         let this = self.eval_context_mut();
 
         if !this.machine.communicate {
-            throw_unsup_format!("`open` not available when isolation is enabled")
+            throw_unsup_format!("`read` not available when isolation is enabled")
         }
 
         let tcx = &{ this.tcx.tcx };
 
         let fd = this.read_scalar(fd_op)?.to_i32()?;
-        let buf = this.force_ptr(this.read_scalar(buf_op)?.not_undef()?)?;
+        let buf_scalar = this.read_scalar(buf_op)?.not_undef()?;
         let count = this.read_scalar(count_op)?.to_usize(&*this.tcx)?;
 
         // Remove the file handle to avoid borrowing issues
-        this.remove_handle_and(
-            fd,
-            |mut handle, this| {
-                let bytes = handle
-                    .file
-                    .read(this.memory_mut().get_mut(buf.alloc_id)?.get_bytes_mut(
-                        tcx,
-                        buf,
-                        Size::from_bytes(count),
-                    )?)
-                    .map(|bytes| bytes as i64);
-                // Reinsert the file handle
-                this.machine.file_handler.handles.insert(fd, handle);
-                this.consume_result(bytes)
-            },
-        )
+        this.remove_handle_and(fd, |mut handle, this| {
+            // Don't use `?` to avoid returning before reinserting the handle
+            let bytes =
+            if count == 0 {
+                Ok(handle.file.read(&mut []))
+            } else {
+                this.force_ptr(buf_scalar).and_then(|buf| this
+                    .memory_mut()
+                    .get_mut(buf.alloc_id).and_then(|alloc|
```

Comment:
this one can be `?` since it's the same error type

Reply:
yes but shoudn't that just return earlier causing this to not reinsert the handle?

Reply:
it's inside the closure, it will only early return the closure

Reply:
Ohh right, I'm missing the parentheses everywhere